crates/gen-host-js/src/lib.rs

```diff
@@ -418,69 +417,91 @@ impl Js {
             Intrinsic::F64ToI64 => self.src.js("
                 function f64ToI64(f) {
-                    I64_TO_F64_F[0] = f;
-                    return I64_TO_F64_I[0];
+                    i64ToF64F[0] = f;
+                    return i64ToF64I[0];
                 }
             "),
 
             Intrinsic::Utf8Decoder => self
                 .src
-                .js("const UTF8_DECODER = new TextDecoder('utf-8');\n"),
+                .js("const utf8Decoder = new TextDecoder('utf-8');\n"),
 
-            Intrinsic::Utf8EncodedLen => self.src.js("let UTF8_ENCODED_LEN = 0;\n"),
+            Intrinsic::Utf16Decoder => self
+                .src
+                .js("const utf16Decoder = new TextDecoder('utf-16');\n"),
+
+            Intrinsic::Utf8EncodedLen => self.src.js("let utf8EncodedLen = 0;\n"),
 
             Intrinsic::Utf8Encode => self.src.js("
-                const UTF8_ENCODER = new TextEncoder('utf-8');
+                const utf8Encoder = new TextEncoder('utf-8');
 
-                function utf8_encode(s, realloc, memory) {
+                function utf8Encode(s, realloc, memory) {
                     if (typeof s !== 'string') \
                         throw new TypeError('expected a string');
 
                     if (s.length === 0) {
-                        UTF8_ENCODED_LEN = 0;
+                        utf8EncodedLen = 0;
                         return 1;
                     }
 
-                    let alloc_len = 0;
+                    let allocLen = 0;
                     let ptr = 0;
                     let writtenTotal = 0;
                     while (s.length > 0) {
-                        ptr = realloc(ptr, alloc_len, 1, alloc_len + s.length);
-                        alloc_len += s.length;
-                        const { read, written } = UTF8_ENCODER.encodeInto(
+                        ptr = realloc(ptr, allocLen, 1, allocLen + s.length);
+                        allocLen += s.length;
+                        const { read, written } = utf8Encoder.encodeInto(
                             s,
-                            new Uint8Array(memory.buffer, ptr + writtenTotal, alloc_len - writtenTotal),
+                            new Uint8Array(memory.buffer, ptr + writtenTotal, allocLen - writtenTotal),
                         );
                         writtenTotal += written;
                         s = s.slice(read);
                     }
-                    if (alloc_len > writtenTotal)
-                        ptr = realloc(ptr, alloc_len, 1, writtenTotal);
-                    UTF8_ENCODED_LEN = writtenTotal;
+                    if (allocLen > writtenTotal)
+                        ptr = realloc(ptr, allocLen, 1, writtenTotal);
+                    utf8EncodedLen = writtenTotal;
+                    return ptr;
+                }
+            "),
+
+            Intrinsic::Utf16Encode => self.src.js("
+                const isLE = new Uint8Array(new Uint16Array([1]).buffer)[0] === 1;
```

Comment:
Handling endianness here is interesting since in general the JS generator doesn't do great with endianness today. Handing out typed array views of wasm content currently always hands out a little-endian view of memory and doesn't try to account for endianness. Single-value loads and stores, however, handle endianness at this time although that was because `DataView` provided the flag and the only reason `DataView` was used historically was for unaligned accesses which are now traps in the component model so we should probably replace `DataView` with typed arrays as well.

In any case as-written this PR I think handles endianness during encoding here but not during decoding. I'm not sure how well supported the `utf-16le` encoding is but that could possibly be an easy fix for the decoding portion.

Do you have a need to support little-endian at this time? If not I'd be tempted to basically `assert(isLE)` at the top of the generated script and defer to https://github.com/bytecodealliance/wit-bindgen/issues/53 for fixing the underlying issue

Reply:
Oh I definitely think that endianness needs to be handled correctly, but I mostly wanted to point out that the handling here is only one piece of the picture and there's more places. The decoding was one but I wanted to point out the preexisting locations where it's not handled as well. I believe the preexisting locations are in `ListCanon{Lower,Lift}` where we're reading a typed array from wasm (`Lift`) and copying a host-defined typed array into wasm (`Lower`)

Reply:
Understood, can we keep the scope of this PR to UTF16 handling, while accepting partial endian support? I've separated out the endian detection intrinsic.